skyd/api.go

```diff
@@ -0,0 +1,180 @@
+
+	"github.com/SkynetLabs/blocker/database"
+	"github.com/sirupsen/logrus"
+	"gitlab.com/NebulousLabs/errors"
+)
+
+const (
+	// skydTimeout is the timeout of the http calls to skyd in seconds
+	skydTimeout = "30"
+)
+
+// SkydAPI is a helper struct that exposes some methods that allow making skyd
+// API calls used by both the API and the blocker
+type SkydAPI struct {
+	staticSkydHost        string
+	staticSkydPort        int
+	staticSkydAPIPassword string
+
+	staticDB     *database.DB
+	staticLogger *logrus.Logger
+}
+
+// NewSkydAPI creates a new Skyd API instance.
+func NewSkydAPI(skydHost, skydPassword string, skydPort int, db *database.DB, logger *logrus.Logger) (*SkydAPI, error) {
+	if db == nil {
+		return nil, errors.New("no DB provided")
+	}
+	if logger == nil {
+		return nil, errors.New("no logger provided")
+	}
+
+	return &SkydAPI{
+		staticSkydHost:        skydHost,
+		staticSkydPort:        skydPort,
+		staticSkydAPIPassword: skydPassword,
+
+		staticDB:     db,
+		staticLogger: logger,
+	}, nil
+}
+
+// BlockSkylinks will perform an API call to skyd to block the given skylinks
+func (skyd *SkydAPI) BlockSkylinks(sls []string) error {
+	// Build the call to skyd.
+	reqBody := skyapi.SkynetBlocklistPOST{
+		Add:    sls,
+		Remove: nil,
+		IsHash: false,
+	}
+	reqBodyBytes, err := json.Marshal(reqBody)
+	if err != nil {
+		return errors.AddContext(err, "failed to build request body")
+	}
+
+	url := fmt.Sprintf("http://%s:%d/skynet/blocklist?timeout=%s", skyd.staticSkydHost, skyd.staticSkydPort, skydTimeout)
+
+	skyd.staticLogger.Debugf("blockSkylinks: POST on %+s", url)
+	req, err := http.NewRequest(http.MethodPost, url, bytes.NewBuffer(reqBodyBytes))
+	if err != nil {
+		return errors.AddContext(err, "failed to build request to skyd")
+	}
+	req.Header.Set("User-Agent", "Sia-Agent")
+	req.Header.Set("Authorization", skyd.staticAuthHeader())
+
+	skyd.staticLogger.Debugf("blockSkylinks: headers: %+v", req.Header)
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return errors.AddContext(err, "failed to make request to skyd")
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusNoContent && resp.StatusCode != http.StatusOK {
+		respBody, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			skyd.staticLogger.Warn(errors.AddContext(err, "failed to parse response body after a failed call to skyd").Error())
+			respBody = []byte{}
+		}
+		err = errors.New(fmt.Sprintf("call to skyd failed with status '%s' and response '%s'", resp.Status, string(respBody)))
+		skyd.staticLogger.Warn(err.Error())
+		return err
+	}
+	return nil
+}
+
+// IsAllowListed will resolve the given skylink and verify it against the
+// allow list, it returns true if the skylink is present on the allow list
+func (skyd *SkydAPI) IsAllowListed(ctx context.Context, skylink string) bool {
+	// build the request to resolve the skylink with skyd
+	url := fmt.Sprintf("http://%s:%d/skynet/resolve/%s", skyd.staticSkydHost, skyd.staticSkydPort)
```

Comment:
F/U: We can check if it's a V2 skylink before making a request.

Also, resolving the skylink can make a nice helper func.